### PR TITLE
Use std::unordered_set instead of set in blockfilter interface

### PR DESCRIPTION
**Pull request by jimpo**
Use `std::unordered_set` (hash set) instead of `std::set` (tree set) in blockfilter interface, as suggested by @ryanofsky in #12254. This may result in a very minor speedup, but I haven't measured.

This moves `CSipHasher` to it's own file `crypto/siphash.h`, so that it can be used in the libbitcoin_util library without including `hash.{h,cpp}`. I'm open to other suggestions on solving this issue if people would prefer to leave CSipHasher where it is.

### CODE REVIEW

File: src/util.h [hunk 1]
```diff
@@ -346,6 +346,21 @@ std::string CopyrightHolders(const std::string& strPrefix);
...
+ */
+class ByteVectorHash
+{
+private:
+    uint64_t m_k0, m_k1;
+
+public:
+    ByteVectorHash();
+    size_t operator()(const std::vector<unsigned char>& input) const;
+};
```
Comment (laanwj) [inline src/util.h]: I don't think `util.h` is the best place for a generic utility data structure such as this, it's more for assorted operating system functions

File: src/crypto/siphash.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2016-2018 The Bitcoin Core developers
```
Comment (laanwj) [inline src/crypto/siphash.h]: ACK on moving siphash to a separate unit

File: src/util.h (hunk 1 above)
Comment (jimpo) [reply src/util.h]: @laanwj What would be the right place? A new file `utiltypes.{h,cpp}`? Or `util/types.{h,cpp}`?

Comment (laanwj) [reply src/util.h]: or maybe `support/bytevectorhash.h` ?
unless there's a strong need to group this with other things

**Comment by laanwj** [comment]
utACK f4608359e4e643c6e197df822e03226146e37a49
verified move-onlyness of extraction commit

File: src/support/bytevectorhash.cpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <crypto/siphash.h>
+#include <support/bytevectorhash.h>
+#include <random.h>
```
Comment (practicalswift) [inline src/support/bytevectorhash.cpp]:
```
src/support/bytevectorhash.cpp:6:1: warning: #includes are not sorted properly [llvm-include-order]
```

File: src/crypto/siphash.cpp [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+} while (0)
+
+CSipHasher::CSipHasher(uint64_t k0, uint64_t k1)
+{
+    v[0] = 0x736f6d6570736575ULL ^ k0;
+    v[1] = 0x646f72616e646f6dULL ^ k1;
+    v[2] = 0x6c7967656e657261ULL ^ k0;
+    v[3] = 0x7465646279746573ULL ^ k1;
+    count = 0;
+    tmp = 0;
```
Comment (practicalswift) [inline src/crypto/siphash.cpp]: Initialize `count` and `tmp` using default member initializers or in the member initializer list of the constructor?

Comment (jimpo) [reply src/crypto/siphash.cpp]: This is a move-only commit. I'd rather not modify the contents of the code here.

Comment (practicalswift) [reply src/crypto/siphash.cpp]: Makes sense!

**Comment by laanwj** [comment]
@ryanofsky can you take a look here please? according to @jimpo this was your suggestion

**Review by promag — COMMENTED** [review]
@jimpo can you improve PR description of the advantages of this change (beside linking @ryanofsky suggestion)?

Do you have numbers to support this change?

Left some nits.

File: src/support/bytevectorhash.h [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+#define BITCOIN_SUPPORT_BYTEVECTORHASH_H
+
+/**
+ * Implementation of Hash named requirement for a byte vector. This may be used the hash function in
+ * std::unordered_set or std::unordered_map over std::vector<unsigned char>. Internally, a random
+ * instance of SipHash-2-4 is used.
+ */
+class ByteVectorHash
+{
+private:
```
[the remark this answers is not in the document]
Comment (jimpo) [reply src/support/bytevectorhash.h]: I think it's more clear to leave it explicit even if the default for `class` is private.

**Review by ryanofsky — APPROVED** [review]
utACK f4608359e4e643c6e197df822e03226146e37a49. Sorry for missing this previously. I confirmed the first commit is move-only, and I think the changes in the second commit should make it easier to use hash maps more places in the future. I left a suggestion below, but also think this change looks good as it is.

File: src/txmempool.h [hunk 6]
```diff
@@ -15,6 +15,7 @@
 
 #include <amount.h>
 #include <coins.h>
+#include <crypto/siphash.h>
```
[the remark this answers is not in the document]
Comment (jimpo) [reply src/txmempool.h]: It's used in SaltedTxidHasher

File: src/support/bytevectorhash.h [hunk 7]
```diff
@@ -0,0 +1,23 @@
+// Copyright (c) 2018 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SUPPORT_BYTEVECTORHASH_H
+#define BITCOIN_SUPPORT_BYTEVECTORHASH_H
+
+/**
+ * Implementation of Hash named requirement for a byte vector. This may be used the hash function in
+ * std::unordered_set or std::unordered_map over std::vector<unsigned char>. Internally, a random
```
[the remark this answers is not in the document]
Comment (jimpo) [reply src/support/bytevectorhash.h]: I updated the comment to say that it works for any types that internally store (or reference) a byte array, but decided not to change the class name.

**Comment by etscrivner** [comment]
utACK cfabca8749a132220e649d4563fb876afc21ceec

**Comment by sipa** [comment]
Just one not comment: I always imagined the things in 'support' as low level features that are independently usable, and find it a bit strange to put something there that depends on `random` and `crypto/siphash`. Perhaps others have a different understanding, as I don't think it was ever written down somewhere what it was for, but I would just put it in `src/`.

**Comment by jimpo** [comment]
@sipa I'd prefer to start creating more directory structure. How would you feel about a `util/` directory instead and we can put it in there?

**Comment by sipa** [comment]
@jimpo Sounds great, but perhaps something to do independently? I'd prefer to not start a new convention, and then have it linger in a half finished state if there's unclarify about how to move forward.

**Review by ryanofsky — COMMENTED** [review]
utACK cfabca8749a132220e649d4563fb876afc21ceec. Only changes are updating ByteVectorHash comment, adding `final` specifier, and tweaking includes.

**Comment by jimpo** [comment]
@sipa I moved bytevectorhash to util/ now.

**Comment by laanwj** [comment]
utACK fef5adcc331c4d7b92b71e03fc8a73343a865599

**Pull request merged**